### PR TITLE
Check and close file if already open

### PR DESCRIPTION
**Pull request by pvyawaha**
*Issue #, if available:*

*Description of changes:* 
Add a check in OTA linux pal and closes file if already open.

By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

### CODE REVIEW

File: platform/posix/ota_pal/source/ota_pal_posix.c [hunk 1]
```diff
@@ -433,17 +434,35 @@ OtaPalStatus_t otaPal_CreateFileForRx( OtaFileContext_t * const C )
             {
                 /* POSIX port using standard library */
                 /* coverity[misra_c_2012_rule_21_6_violation] */
```
Comment (aggarw13) [inline platform/posix/ota_pal/source/ota_pal_posix.c:436]: Does the comment about MISRA violation now need to be moved to above the `fclose` call?

File: platform/posix/ota_pal/source/ota_pal_posix.c [hunk 2]
```diff
@@ -433,17 +434,35 @@ OtaPalStatus_t otaPal_CreateFileForRx( OtaFileContext_t * const C )
...
-                C->pFile = fopen( ( const char * ) realFilePath, "w+b" );
-
                 if( C->pFile != NULL )
```
Comment (aggarw13) [inline platform/posix/ota_pal/source/ota_pal_posix.c:437]: A comment would be good to mention about this logic closing a file if already open.

File: platform/posix/ota_pal/source/ota_pal_posix.c [hunk 3]
```diff
@@ -433,17 +434,35 @@ OtaPalStatus_t otaPal_CreateFileForRx( OtaFileContext_t * const C )
...
-                    result = OTA_PAL_COMBINE_ERR( OtaPalSuccess, 0 );
-                    LogInfo( ( "Receive file created." ) );
+                    filerc = fclose( C->pFile );
+
+                    if( filerc != 0 )
+                    {
+                        LogError( ( "Failed to close OTA update file." ) );
+                        result = OTA_PAL_COMBINE_ERR( OtaPalFileClose, errno );
+                    }
+                    else
```
Comment (aggarw13) [inline platform/posix/ota_pal/source/ota_pal_posix.c:446]: (Nit) `filerc` can be avoided
```suggestion
                    if( fclose( C->pFile ) != 0)
                    {
                        LogError( ( "Failed to close OTA update file." ) );
                        result = OTA_PAL_COMBINE_ERR( OtaPalFileClose, errno );
                    }
                    else
```

File: platform/posix/ota_pal/source/ota_pal_posix.c [hunk 4]
```diff
@@ -433,17 +434,35 @@ OtaPalStatus_t otaPal_CreateFileForRx( OtaFileContext_t * const C )
...
+
+                    if( C->pFile != NULL )
+                    {
+                        result = OTA_PAL_COMBINE_ERR( OtaPalSuccess, 0 );
+                        LogInfo( ( "Receive file created." ) );
+                    }
+                    else
+                    {
+                        result = OTA_PAL_COMBINE_ERR( OtaPalRxFileCreateFailed, errno );
+                        LogError( ( "Failed to start operation: Operation already started. failed to open -- %s Path ", C->pFilePath ) );
```
Comment (aggarw13) [inline platform/posix/ota_pal/source/ota_pal_posix.c:464]: To make this log more useful in the context of viewing a set log messages on the console, mentioning which specific specific failed will be good avoid the "Which operation failed" question that might arise to a person debugging.

Comment (pvyawaha) [reply platform/posix/ota_pal/source/ota_pal_posix.c:465]: Added in https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1551/commits/b99d4ab575f455287a26210bb5d10cf318aa3d03

File: platform/posix/ota_pal/source/ota_pal_posix.c (hunk 2 above)
Comment (pvyawaha) [reply platform/posix/ota_pal/source/ota_pal_posix.c:436]: Added in https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1551/commits/b99d4ab575f455287a26210bb5d10cf318aa3d03

File: platform/posix/ota_pal/source/ota_pal_posix.c (hunk 1 above)
Comment (pvyawaha) [reply platform/posix/ota_pal/source/ota_pal_posix.c]: Added in https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1551/commits/b99d4ab575f455287a26210bb5d10cf318aa3d03

File: platform/posix/ota_pal/source/ota_pal_posix.c (hunk 4 above)
Comment (pvyawaha) [reply platform/posix/ota_pal/source/ota_pal_posix.c]: Addressed in https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1551/commits/2b7ac96e596b688e77b53f7fa291dce7cc1d0893

File: platform/posix/ota_pal/source/ota_pal_posix.c [hunk 5]
```diff
@@ -410,58 +410,66 @@ OtaPalStatus_t otaPal_Abort( OtaFileContext_t * const C )
...
+    assert( C->pFilePath != NULL );
+
+    if( C->pFilePath[ 0 ] != ( uint8_t ) '/' )
+    {
+        status = getFilePathFromCWD( realFilePath, ( const char * ) C->pFilePath );
+    }
+    else
     {
-        if( C->pFilePath != NULL )
+        ( void ) strncpy( realFilePath, ( const char * ) C->pFilePath, strlen( ( const char * ) C->pFilePath ) + 1U );
```
Comment (aggarw13) [inline platform/posix/ota_pal/source/ota_pal_posix.c:430]: A comment about the reason for `+ 1U` would be good to add.

File: platform/posix/ota_pal/source/ota_pal_posix.c [hunk 6]
```diff
@@ -410,58 +410,66 @@ OtaPalStatus_t otaPal_Abort( OtaFileContext_t * const C )
...
+            {
+                result = OTA_PAL_COMBINE_ERR( OtaPalSuccess, 0 );
+                LogInfo( ( "Receive file created." ) );
             }
             else
             {
-                LogError( ( "Could not generate the absolute path for the file" ) );
-                result = OTA_PAL_COMBINE_ERR( OtaPalRxFileCreateFailed, 0 );
+                result = OTA_PAL_COMBINE_ERR( OtaPalRxFileCreateFailed, errno );
+                LogError( ( "Failed to open file for update-- %s Path ", C->pFilePath ) );
```
Comment (aggarw13) [inline platform/posix/ota_pal/source/ota_pal_posix.c:465]:
```suggestion
                LogError( ( "Failed to open file for OTA update. Path: %s ", C->pFilePath ) );
```

File: platform/posix/ota_pal/source/ota_pal_posix.c [hunk 7]
```diff
@@ -410,58 +410,66 @@ OtaPalStatus_t otaPal_Abort( OtaFileContext_t * const C )
...
         }
-        else
-        {
-            result = OTA_PAL_COMBINE_ERR( OtaPalRxFileCreateFailed, 0 );
-            LogError( ( "Invalid file path provided." ) );
-        }
     }
     else
     {
+        LogError( ( "Could not generate the absolute path for the file" ) );
```
Comment (aggarw13) [inline platform/posix/ota_pal/source/ota_pal_posix.c:471]:
```suggestion
        LogError( ( "Could not generate the absolute path for the receive file" ) );
```

File: platform/posix/ota_pal/source/ota_pal_posix.c (hunk 5 above)
Comment (yanjos-dev) [reply platform/posix/ota_pal/source/ota_pal_posix.c:431]: done

File: platform/posix/ota_pal/source/ota_pal_posix.c [hunk 8]
```diff
@@ -410,58 +410,67 @@ OtaPalStatus_t otaPal_Abort( OtaFileContext_t * const C )
...
+        if( C->pFile != NULL )
         {
-            if( C->pFilePath[ 0 ] != ( uint8_t ) '/' )
+            /* POSIX port using standard library */
+            /* coverity[misra_c_2012_rule_21_6_violation] */
+            if( fclose( C->pFile ) != 0 )
             {
-                status = getFilePathFromCWD( realFilePath, ( const char * ) C->pFilePath );
+                LogError( ( "Failed to close OTA update file." ) );
+                result = OTA_PAL_COMBINE_ERR( OtaPalFileClose, errno );
```
Comment (aggarw13) [inline platform/posix/ota_pal/source/ota_pal_posix.c:444]: Does `OtaPalFileClose` represent a failure in closing the file?

Comment (yanjos-dev) [reply platform/posix/ota_pal/source/ota_pal_posix.c:444]: Yes, you can find the description of this here:
https://github.com/aws/ota-for-aws-iot-embedded-sdk/blob/bfc47ed79bf7cc47bc70860c0f471f67f34ed79c/source/include/ota_platform_interface.h#L89

Comment (aggarw13) [reply platform/posix/ota_pal/source/ota_pal_posix.c:444]: Thanks for sharing the link. Imo, naming the enumeration value as `OtaFileCloseFailure` (and similarly for the `OtaFileAbort`) would be more expressive.

**Review by aggarw13 — APPROVED** [review]

**Review by aggarw13 — APPROVED** [review]

**Pull request closed without merging**